Sync With Upstream

l-monninger opened the pull request:
Resolves #41

### Description

### Test Plan
<!-- Please provide us with clear details for verifying that your changes work. -->

0xmovses committed:
update zeroize

0xmovses committed:
Update Cargo.toml

0xmovses committed:
Update Cargo.toml

0xmovses committed:
Update Cargo.toml

0xmovses committed:
Update Cargo.toml

0xmovses committed:
add into

0xmovses committed:
more into

0xmovses committed:
update visibility

0xmovses committed:
derive serde for ValidatorSigner

0xmovses committed:
Merge pull request #1 from 0xmovses/main

Monza

l-monninger committed:
feat: monza-testing.

l-monninger committed:
fix: pub api struct.

0xmovses committed:
Merge pull request #2 from 0xmovses/devnet

Devnet merge

l-monninger committed:
feat: pub modules.

l-monninger committed:
fix: skipped patches.

0xmovses committed:
make serives pub

0xmovses committed:
Merge pull request #3 from movementlabsxyz/0xmovses/pub-services

Make `services` pub

0xmovses committed:
fix dalek

0xmovses committed:
pub mod network

0xmovses committed:
pub typo

0xmovses committed:
add error logs for deserialization

0xmovses committed:
pub mod core_mempool

mzabaluev committed:
Don't version control VSCode settings

Let developers configure their VSCode without
turning this into commit fests and battles of preferences.

mzabaluev committed:
types: make ring an optional dependency

Introduce the "testing" feature to gate the test_utils module
for external use. This feature also gates the ring dependency.

mzabaluev committed:
aptos-crypto: move noise support to aptos-network

Apart from aptos-telemetry which we should not need,
aptos-network is the only functionality that uses Noise and,
therefore, currently needs the dependency on the ring crate.

0xmovses committed:
initial

0xmovses committed:
add prunes

0xmovses committed:
comment

mzabaluev committed:
consensus-types: remove request_response

Eliminate the async dependency in consensus-types.

l-monninger committed:
fix: merge.

l-monninger committed:
Merge pull request #4 from mzabaluev/mikhail/de-scm-vscode-config

chore: remove VSCode configuration from git

0xmovses committed:
add revert_kv method

mzabaluev committed:
cut out tokio from consensus-types

Stub out the status field in ProofWithData with the vector that's
evenutally supposed to be retrieved, until we figure out if we need
the field and the whole structure at all.

0xmovses committed:
add test

0xmovses committed:
update test

mzabaluev committed:
aptos-vm: cut out use of futures

0xmovses committed:
update tests

0xmovses committed:
on reversion, version stays the same

0xmovses committed:
add error assertion test

mzabaluev committed:
aptos-logger: cut out async logging

mzabaluev committed:
schemadb: remove API variability

Use a rocksdb type that is not dependent on the features with which
the crate is compiled, since some code in this crate uses API that's
specific to a particular target type of the DB alias.

0xmovses committed:
Merge pull request #8 from movementlabsxyz/mikhail/fix-schemadb-api-variability

schemadb: Remove API variability

mzabaluev committed:
schemadb: Remove API variability (take 2)

mzabaluev committed:
schemadb: fix up type of DB iterator

0xmovses committed:
Merge pull request #9 from movementlabsxyz/mikhail/fix-schemadb-api-variability

schemadb: Remove API variability (take 2)

l-monninger committed:
feat: mempool visibility.

l-monninger committed:
Merge branch 'monza' of https://github.com/movementlabsxyz/aptos-core into monza

mzabaluev committed:
Remove aptos-telemetry

Telemetry uses the noise API from aptos-crypto for its own
handshake implementation, which we don't want to care about
at this point.
Also cut out all dependent crates.

mzabaluev committed:
aptos-crypto: remove mentions of noise

From README

0xmovses committed:
Merge pull request #5 from mzabaluev/mikhail/de-ring

ENG-179 Remove `ring` dependency from aptos-crypto and pare down dependent code

0xmovses committed:
:Merge branch 'monza' into 0xmovses/ENG-206-implement-block-reversion

mzabaluev committed:
Cut out aptos-consensus and dependent crates

aptos-consensus had code that provided consensus-types data
in asynchronous sends. Since we are not likely to need this crate
anyway, cut it out of the workspace. Also remove any dependent crates.

mzabaluev committed:
aptos-network: fix up noise imports in tests

l-monninger committed:
fix: visibility for testing.

l-monninger committed:
Merge branch 'monza' of https://github.com/movementlabsxyz/aptos-core into monza

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: playing with deadlock issues.

l-monninger committed:
fix: cleanup.

0xmovses committed:
revert nth commit

0xmovses committed:
use with_block_nums

0xmovses committed:
make better assertions

0xmovses committed:
stop last test flakiness

l-monninger committed:
fix: make get_transaction_by_hash_inner public.

l-monninger committed:
fix: debugging transaction_by hash issue.

l-monninger committed:
fix: debugging transaction_by hash issue.

l-monninger committed:
fix: debugging transaction_by hash issue.

0xmovses committed:
debugging

l-monninger committed:
fix: debugging transaction_by hash issue.

0xmovses committed:
fix bug

0xmovses committed:
fmt

mzabaluev committed:
aptos-dkg: relax dep on aptos-experimental-runtimes

It's only used in test code, move it to dev-dependencies

mzabaluev committed:
experimiental/runtimes: remove tokio dep

l-monninger committed:
Merge pull request #10 from movementlabsxyz/mikhail/trim-aptos-logger

ENG-251 aptos-logger: cut out async logging

0xmovses committed:
Merge branch 'monza' into 0xmovses/ENG-206-implement-block-reversion

0xmovses committed:
add reversion to Writer trait, enable epoch reversion

0xmovses committed:
add new commit reversion method

l-monninger committed:
Merge pull request #7 from movementlabsxyz/mikhail/de-async

ENG-180: Cut out async dependencies from consensus-types and async-vm

l-monninger committed:
Merge pull request #6 from movementlabsxyz/0xmovses/ENG-206-implement-block-reversion

ENG-206 Implement Block Reversion

Commit:
chore: remove logging.

Commit:
fix: trying miscellaneous state view id.

mzabaluev committed:
Remove calls to AptosDataBuilder::is_async

Some tests still had these.

mzabaluev committed:
aptos-db: remove reversion helper method

AptosDB::commit_reversion_ledger_info
does the same thing as commit_ledger_info,
just use that method instead.

mzabaluev committed:
storage-interface: add FinalityView

This wrapper is meant to represent the finalized state
for the API.

mzabaluev committed:
storage-interface: tests for FinalityView

mzabaluev committed:
storage-interface: blanket impl DbReader for Arc

0xmovses committed:
update set ledger info method, update tests

0xmovses committed:
update tests update set method

mzabaluev committed:
storage-interface: fixed up finality_view tests

No randomness allows us to verify the hash.

mzabaluev committed:
storage-interface: fix DbReader impl for Arc

get_read_delegatee needs to deref to get the vtable
for the pointee type.

mzabaluev committed:
storage-interface: rename setter for FinalityView

FinalityView::set_finalized_block_height better describes
the argument.

mzabaluev committed:
storage-interface: remove a dbg printout

mzabaluev committed:
Revert "fix: debugging transaction_by hash issue."

This reverts commit 7f3e0e5c93d2b208f421445e3fd7ecdad8d0a4dd.

mzabaluev committed:
storage-interface: finality_view tests ret Result

Use ? for terser, idiomatic code that can be cribbed for production.

mzabaluev committed:
storage-interface: latest state vew test

Draft a test for FinalityView, checking that the latest_state_checkpoint_view
extension method gets us the view into finalized state.
StateView support in the mock is needed to complete the test.

mzabaluev committed:
storage-interface: clean up and comment

In FinalityView::set_finalized_block_height, remove the outdated
sanity check, add comments explaining why we can use the empty hash
and no signatures specifically for Movement.

mzabaluev committed:
storage: improve FinalityView set method

For set_finalized_block_height, retrieve ledger info from the
corresponding epoch or use the latest.

mzabaluev committed:
storage: more FinalityView testing

Exercise the DbStateView instance returned from
latest_state_checkpoint_view. No testing yet if the state
matches expectations for the finalized version.

mzabaluev committed:
Revert "storage: improve FinalityView set method"

This reverts commit 1731cce36e3810d4034240deeefffd3745f8dedc.

mzabaluev committed:
api: expose AccountsApi::get_account_inner

Access to account state is convenient to test with.

mzabaluev committed:
Merge pull request #13 from movementlabsxyz/mikhail/finalized-state

`FinalityView`, a `DbReader` wrapper to represent a finalized state

l-monninger committed:
chore: api_connection_config should be public.

l-monninger committed:
fix: pin diesel.

l-monninger committed:
fix: pin diesel.

l-monninger committed:
fix: checking write table output.

l-monninger committed:
Merge pull request #14 from movementlabsxyz/l-monninger/faucet-instrumentation

Faucet Instrumentation

mzabaluev committed:
aptos-db: revert_commit improvements

Pass LedgerInfoWithSignatures by reference like in other methods.
Do not try to set epoch and version on the LedgerInfo in a way that
does not work. Instead, it's expected to be correct for the latest
pre-revert version.

mzabaluev committed:
aptos-db: simplify signature of revert_commit

All necessary information can be obtained from
either the LedgerInfoWithSignatures, or the current
database state.

mzabaluev committed:
aptos-db: fix excessive batching

Update both progress values in the ledger metadata
schema in one batch.

mzabaluev committed:
Merge pull request #22 from movementlabsxyz/mikhail/ledger-info-on-revert

Make `LedgerInfo` the source of truth on revert

mzabaluev committed:
vm-genesis: prevent division by zero

Rewrite the computation for rewards rate to avoid
division by zero when the epoch duration exceeds one year.

mzabaluev committed:
Merge pull request #23 from movementlabsxyz/mikhail/fix-zero-divide-on-long-epochs

fix: prevent zero divide on long epoch duration

mzabaluev committed:
fix: cherry-pick.

mzabaluev committed:
aptos-db: batching fixes in revert_commit

Use batches correctly: one per schema, commit all batches.

mzabaluev committed:
aptos-db: reorder LedgerInfo commit on revert

Commit (and validate) the LedgerInfoWithSignatures
and update the DbMetadataKey::OverallCommitProgress value
first thing when reverting commits.

mzabaluev committed:
fix: cherry-pick.

mzabaluev committed:
aptos-db: remove more on revert_commit

Remove block information from ledger metadata db,
update the LedgerCommitProgress marker in the same atomic batch.
In StateKvDb, reuse the commit method to update version markers.

Stub out reversal of store indices from StateStore, add TODO comments
on what will need to be cleaned up.

Consistently reuse the prune functionality available for
the relevant tables to remove the ranges of versions. Remove
methods added earlier that
Fix off-by-one errors with version ranges to prune.

mzabaluev committed:
fix: cherry-pick

mzabaluev committed:
aptos-db: remove unused imports in tests

mzabaluev committed:
fix: merge.

l-monninger committed:
fix: merge.

l-monninger committed:
fix: aptos-telemetry.

l-monninger committed:
fix: restoring aptos-admin-service.

l-monninger committed:
fix: aptos-inspection-service.

l-monninger committed:
fix: restore aptos-debugger.

l-monninger committed:
fix: grpc cache workers.

l-monninger committed:
fix: indexer-grpc-file-store.

l-monninger committed:
fix: indexer-grpc-integration-tests.

l-monninger committed:
fix: aptos-debugger.

l-monninger committed:
fix: aptos-forge.

l-monninger committed:
fix: aptos-forge.

l-monninger committed:
fix: aptos-release-builder.

l-monninger committed:
fix: aptos-rosetta.

l-monninger committed:
feat(fin): update to match DbReader changes.

mzabaluev committed:
aptos-db: set in-memory ledger info on revert

l-monninger committed:
fix: revert commit.

l-monninger committed:
fix: revert commit.

l-monninger committed:
fix: bad argument.

mzabaluev committed:
fix: cherry-pick.

mzabaluev commented:
(The diff view in GitHub is unusable, will post excerpts)

We used to need this impl for `FinalityView`: 

```diff
--- a/storage/storage-interface/src/lib.rs
+++ b/storage/storage-interface/src/lib.rs
@@ -479,58 +485,11 @@ pub trait DbReader: Send + Sync {
         state_key: &StateKey,
         version: Version,
     ) -> Result<(Option<StateValue>, SparseMerkleProof)> {
-        self.get_state_value_with_proof_by_version_ext(state_key, version)
+        self.get_state_value_with_proof_by_version_ext(state_key, version, 0)
             .map(|(value, proof_ext)| (value, proof_ext.into()))
     }
 }
 
-impl DbReader for Arc<dyn DbReader> {
-    fn get_read_delegatee(&self) -> &dyn DbReader {
-        &**self
-    }
-}
```

Otherwise, the changes look OK.

mzabaluev commented on aptos-move/e2e-move-tests/Cargo.toml:
```diff
@@ -22,7 +22,7 @@ aptos-gas-profiling = { workspace = true }
 aptos-gas-schedule = { workspace = true, features = ["testing"] }
 aptos-language-e2e-tests = { workspace = true }
 aptos-package-builder = { workspace = true }
-aptos-types = { workspace = true }
```
Do we really need testing features in production code?

The "testing" feature requirement should be moved to the dev-dependencies section.

mzabaluev commented on aptos-move/vm-genesis/Cargo.toml:
```diff
@@ -17,7 +17,7 @@ aptos-cached-packages =  { workspace = true }
 aptos-crypto = { workspace = true }
 aptos-framework =  { workspace = true }
 aptos-gas-schedule = { workspace = true } 
-aptos-types = { workspace = true }
```
Same as above.

mzabaluev commented on config/src/config/identity_config.rs:
```diff
@@ -38,7 +38,15 @@ pub struct IdentityBlob {
 impl IdentityBlob {
     pub fn from_file(path: &Path) -> anyhow::Result<IdentityBlob> {
-        Ok(serde_yaml::from_str(&fs::read_to_string(path)?)?)
+        let content = fs::read_to_string(path)?;
+        match serde_yaml::from_str(&content) {
+            Ok(identity_blob) => Ok(identity_blob),
+            Err(e) => {
+                eprintln!("Deserialization error: {:?}", e);
+                eprintln!("YAML content:\n{}", content);
+                Err(e.into())
+            },
+        }
```
Looks like temporary debug code, could be replaced with upstream.

mzabaluev commented on crates/indexer/src/models/move_tables.rs:
```diff
@@ -56,6 +57,10 @@ impl TableItem {
         transaction_version: i64,
         transaction_block_height: i64,
     ) -> (Self, CurrentTableItem) {
+        debug!(
+            "Decoding WriteTableItem: {:?}",
+            write_table_item
+        );
```
Debug code, could be removed from the delta.

mzabaluev commented on execution/executor/src/block_executor.rs:
```diff
@@ -229,7 +229,7 @@ where
                         .start_timer();
                     info!("next_version: {}", parent_output.next_version());
                     CachedStateView::new(
-                        StateViewId::BlockExecution { block_id },
```
This needs to be in as our workaround, right?

mzabaluev reviewed (approved):
The rebasing of our changes looks sensible.

Please switch the base to `monza` before merging.

mzabaluev commented on aptos-move/e2e-move-tests/Cargo.toml:
```diff
@@ -22,7 +22,7 @@ aptos-gas-profiling = { workspace = true }
 aptos-language-e2e-tests = { workspace = true }
 aptos-package-builder = { workspace = true }
-aptos-types = { workspace = true }
+aptos-types = { workspace = true, features = ["testing"] }
```
Do we really need testing features in production code?

The "testing" feature requirement should be moved to the dev-dependencies section.

mzabaluev commented on aptos-move/vm-genesis/Cargo.toml:
```diff
@@ -17,7 +17,7 @@ aptos-cached-packages =  { workspace = true }
 aptos-framework =  { workspace = true }
 aptos-gas-schedule = { workspace = true } 
-aptos-types = { workspace = true }
+aptos-types = { workspace = true, features = ["testing"] }
```
Same as above.

mzabaluev replied:
This is needed, it's too cumbersome to extract a subset used in vm-genesis, and it's some lazy-parsed JSON.
So let's keep this as a visible wart.

mzabaluev committed:
types: remove setter methods

These were added to make LedgerInfo and BlockInfo work
when reverting, but are no longer needed after rewrites.

mzabaluev committed:
aptos-db: truncate state merkle DB on revert

mzabaluev committed:
aptos-db: remove an unused test helper

mzabaluev committed:
fix: cherry-pick.

mzabaluev committed:
aptos-db: ResetLock

Add a ResetLock to expose scoped locking used in `reset`,
to implement atomic updates of buffered state in revert_commit.

mzabaluev committed:
aptos-db: Use the state store lock on revert

l-monninger committed:
fix: merge.

l-monninger committed:
Merge pull request #40 from movementlabsxyz/l-monninger/sync-with-upstream-helper

A Helper for Syncing w/ the Upstream

mzabaluev commented on execution/executor/src/block_executor.rs:
```diff
@@ -229,7 +229,7 @@ where
                     info!("next_version: {}", parent_output.next_version());
                     CachedStateView::new(
-                        StateViewId::BlockExecution { block_id },
+                        StateViewId::Miscellaneous,
```
Does this need to remain as our workaround?

mzabaluev commented on storage/aptosdb/src/db/include/aptosdb_writer.rs:
```diff
@@ -197,6 +199,76 @@ impl DbWriter for AptosDB {
     }
+
+    /// Revert a commit.
+    fn revert_commit(&self, ledger_info_with_sigs: &LedgerInfoWithSignatures) -> Result<()> {
```
This is cherry-picked from #28, not yet reviewed there. As I wrote this code, I'll leave it as an exercise for @0xmovses.

mzabaluev commented on storage/aptosdb/src/db/include/aptosdb_writer.rs:
```diff
@@ -561,18 +688,6 @@ impl AptosDB {
                 new_root_hash,
                 expected_root_hash,
             );
-            let current_epoch = self
-                .ledger_db
-                .metadata_db()
-                .get_latest_ledger_info_option()
-                .map_or(0, |li| li.ledger_info().next_block_epoch());
-            ensure!(
-                x.ledger_info().epoch() == current_epoch,
-                "Gap in epoch history. Trying to put in LedgerInfo in epoch: {}, current epoch: {}",
-                x.ledger_info().epoch(),
-                current_epoch,
-            );
```
I'm not sure this check really needs to be removed...

mzabaluev commented on storage/aptosdb/src/db/include/aptosdb_writer.rs:
```diff
@@ -630,4 +745,42 @@ impl AptosDB {
+    // Update in-memory state of the database and the metrics before reverting.
+    // Note that any failures in persisting the revert should be treated as
+    // non-recoverable.
+    fn pre_revert(
```
This is also a cherry-pick from #28, not yet used anywhere, but will be once that PR lands.

mzabaluev commented on storage/aptosdb/src/state_store/mod.rs:
```diff
@@ -549,15 +551,17 @@ impl StateStore {
 
     pub fn reset(&self) {
-        let (buffered_state, smt_ancestors) = Self::create_buffered_state_from_latest_snapshot(
+        let lock = self.reset_lock();
+        lock.reset();
+    }
+
+    pub fn reset_lock(&self) -> ResetLock<'_> {
```
Another cherry-pick from #28; it looks like a needless refactoring now, but will come useful later.

mzabaluev commented on testsuite/forge/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ rust-version = { workspace = true }
 [dependencies]
 again = { workspace = true }
 anyhow = { workspace = true, features = ["backtrace"] }
-aptos = { workspace = true }
```
This is a zombie file resurrected by merge. As the crate does not exist in the movement fork, could be removed.

mzabaluev commented on testsuite/smoke-test/Cargo.toml:
```diff
@@ -14,7 +14,7 @@ rust-version = { workspace = true }
 [dependencies]
 anyhow = { workspace = true }
-aptos = { workspace = true, features = ["fuzzing"] }
+movement = { workspace = true, features = ["fuzzing"] }
```
This is a zombie file resurrected by merge. As the crate does not exist in the movement fork, could be removed.

mzabaluev commented on testsuite/testcases/Cargo.toml:
```diff
@@ -14,7 +14,7 @@ rust-version = { workspace = true }
 [dependencies]
 anyhow = { workspace = true }
-aptos = { workspace = true, features = ["fuzzing"] }
+movement = { workspace = true, features = ["fuzzing"] }
```
This is a zombie file resurrected by merge. As the crate does not exist in the movement fork, could be removed.

mzabaluev commented:
The changes look sensible now, apart from some minor issues. The only real concern that jumped at me is the added `todo!` macros, potentially introducing crashiness.

mzabaluev committed:
movement: revert mod genesis from upstream

Replace todo! in struct value initializers with sensible values.
Apply some typo fixes.

mzabaluev committed:
Remove debug printouts not present upstream

mzabaluev committed:
Remove nft-metadata-crawler-parser again

mzabaluev committed:
aptos-storage-interface: recover tests

mzabaluev committed:
Restore crates from upstream

When syncing with upstream, merge conflicts of the "changed by them/
removed by us" kind have been resolved by updating the conflicting
files to the latest upstream revision. This only partially recovered
the source for crates that have been removed for Monza.
Restore the crates fully.

Some of the changes made for Monza are overwritten from upstream
to reduce the delta and allow resurrecting some crates erased
in the work for Monza:
- ring dependency removal from aptos-crypto, moving the noise
  implementation to aptos-network;
- removal of tokio synchronization slots from aptos-types.

This means this branch no longer supports Monza, and those changes
would need to be re-applied if the Monza direction is taken again.

mzabaluev committed:
aptos-node: recover mod indexer

mzabaluev committed:
aptos-db: update tests

The storage interface used by tests added by Movement
needed to be updated to the upstream changes.

mzabaluev commented:
The merge breaks the `movement` CLI (in `./crates/aptos`), because it updates many of its dependencies, and resurrects others in a broken form, by seemingly resolving "removed by us/changed by them" conflicts in favor of the usually incomplete set of files changed upstream. The conflicts in `Cargo.toml` are also resolved in favor of resurrecting these in

I think the sensible way forward is to completely check out the required crates from `main`, and incorporate #41 into this so that the `movement` CLI is up-to-date against its workspace dependencies. Unfortunately, this means partially overwriting some changes made for Monza. We can keep them in the `monza` branch and revisit if we decide to pick up work on that project again.

mzabaluev committed:
fix(indexer-grpc-server-framework): fixup

Pull the README from upstream and remove the duplicate workspace entry.

l-monninger commented:
My mistake, I misused `--all-targets` I guess: https://github.com/rust-lang/cargo/issues/10958

mzabaluev committed:
Sync movement CLI from aptos-cli-v3.5.0

l-monninger committed:
Merge pull request #43 from movementlabsxyz/mikhail/sync-with-upstream-fixups

Sync With Upstream: fixups, updating the CLI to 3.5.0

mzabaluev commented on crates/aptos/CHANGELOG.md:
```diff
@@ -1,92 +1,52 @@
-## Unreleased
-- Add balance command to easily get account balances for APT currently
-
-## [3.4.1] - 2024/05/31
```
Re-syncing is done in #43.

andygolay replied:
Do you want to rename the associated `aptos` directory to `movement` ?

mzabaluev commented:
This should be good to merge now.

mzabaluev committed:
fix(aptos-schemadb): revert single-threaded mode

The changes do not work if the "multi-threaded-cf" feature
is enabled. We only tried this for Monza.

mzabaluev committed:
Revert "revert single-threaded mode"

This reverts commit bc283e633aac9bb72a5d47b970348d5c60816e0e.

mzabaluev committed:
fix: more feature-agnostic rocksdb API use

mzabaluev committed:
fix(aptos-release-builder): restore from upstream

mzabaluev committed:
fix: rename some aptos imports to movement

mzabaluev committed:
fix(aptos-smoke-test): restore from upstream

mzabaluev committed:
fix(aptos-testcases): rename imports

crate aptos renamed to movement.

mzabaluev committed:
refactor(storage-interface): dyn FinalityView

We only parameterized FinalityView with Arc<dyn DbReader>,
so just bake it in rather than relying on a generic impl.
When DbReaderWriter is refactored, we might get a better shot at
improving this.

mzabaluev committed:
fix: make get_transaction_by_hash_inner public

It was made private again in the merge.
